csr2transmart/csr_mapper.py

```diff
@@ -37,6 +42,96 @@ def row_value_to_value(concept_col, value_type: ValueType) -> Optional[Value]:
         return TextValue(concept_col)
 
 
+class ObservationMapper:
+    """
+    Map observations for subject registry and study registry
+    """
+    def __init__(self,
+                 default_trial_visit: TrialVisit,
+                 individual_id_to_patient: Dict[str, Patient],
+                 concept_key_to_concept: Dict[str, Concept],
+                 concept_key_to_modifier_key: Dict[str, str],
+                 modifier_key_to_modifier: Dict[str, Modifier]):
+        self.default_trial_visit = default_trial_visit
+        self.individual_id_to_patient = individual_id_to_patient
+        self.concept_key_to_concept = concept_key_to_concept
+        self.concept_key_to_modifier_key = concept_key_to_modifier_key
+        self.modifier_key_to_modifier = modifier_key_to_modifier
+        self.observations: List[Observation] = []
+
+    def map_observation_metadata(self, modifier_key, value: str) -> Optional[ObservationMetadata]:
+        if modifier_key is None or value is None:
+            return None
+        mod_metadata: Dict[Modifier, Value] = dict()
+        modifier = self.modifier_key_to_modifier.get(modifier_key)
+        if modifier is None:
+            return None
+        mod_metadata[modifier] = CategoricalValue(value)
+        return ObservationMetadata(mod_metadata)
+
+    def add_observations_for_entity(self, entity, entity_id: str, patient: Patient):
+        concept_keys = entity.fields.keys()
```

Comment:
I think the concept key should be a concatenation of the entity name and the field name, e.g., `Individual.age`. One reason is that we avoid possible naming collisions. Another is that for the duplicate individual ids, we don't have to increase the instance_num to avoid data collisions.

Reply:
And then we can use the concept key instead of `random.choice` for the concept code.